lib/transaction/pending.go

```diff
@@ -0,0 +1,125 @@
+// Implement operations related to pending operations
+//
+// Pending operations are recorded blockchain operations that
+// have not yet altered the state of the node,
+// but will trigger at a specific block height.
+package transaction
+
+import (
+	"container/list"
+	"fmt"
+	"sync"
+)
+
+// The main type is a simple list with insert/pop primitives
+type PendingPool struct {
+	sync.RWMutex
+	*list.List
+}
+
+// Unexported item type
+// We construct it from a height/hash pair,
+// but the user only needs to get the hash back
+// as popping requires the height
+type pendingPoolItem struct {
+	// Block height at which this operation will take action
+	height uint64
+	// Operation hash
+	txHash string
+}
+
+// Instantiate a new `PendingPool`
+func NewPendingPool() *PendingPool {
+	return &PendingPool{sync.RWMutex{}, list.New()}
+}
+
+// Insert a new element in the pool
+func (pp *PendingPool) Insert(height uint64, hash string) {
+	elem := &pendingPoolItem{height: height, txHash: hash}
+
+	pp.Lock()
+	defer pp.Unlock()
+
+	// Otherwise we need to insert this item in order
+	for e := pp.Front(); e != nil; e = e.Next() {
+		casted := e.Value.(*pendingPoolItem)
+		// We can have multiple tx at the same height
+		// Store them in the order they are inserted
+		if casted.height > height {
+			pp.InsertBefore(elem, e)
+			return
+		}
+	}
+	// Either empty, or we need to insert at the end
+	pp.PushBack(elem)
+}
+
+// Peek an element from the list
+//
+// Elements are always stored in order, and should be popped in order.
+// This means it's fine to `Insert` out of order
+// (e.g. `Insert(42, "...")`, `Insert(25, "...")`,
+// but when `Pop`ing, `25` should be `Pop`ed before `42`.
+//
+// Params:
+//   height = The height being processed.
+//            If the head of the list is under this height,
+//            an empty string is returned.
+//   offset = In the event that this list contains more than one
+//            entry for a given height, this will skip `offset`
+//            entries. As such, it should be 0 on the first call.
+func (pp *PendingPool) Peek(height uint64, offset uint64) string {
+	pp.RLock()
+	defer pp.RUnlock()
+
+	nOffset := offset
+	for entry := pp.Front(); entry != nil; entry = entry.Next() {
```

Comment:
Is the `List` better for performance than `map`?
Doesn't it search front to target sequentially?

Reply:
Yes, that's the point. Elements are kept in order.

Reply:
>  Elements are kept in order.

@Geod24 Yes. But it is still O(n) because it does not binary search.